### PR TITLE
Refactoring part 1

### LINKED ISSUE #98: Fix issue with gridscan creation
Currently the positions for gridscans are being written incorrectly. See https://github.com/DiamondLightSource/python-artemis/issues/191

## Acceptance Criteria
* These are written correctly
* There are tests against this

### PR DESCRIPTION
**Pull request by noemifrisina**
- [x] Redefine NXmxWriter 
- [x] Define some utilities for describing Goniometer/Detector/...
- [x] Better VDS slicing
- [x] Add tests ~and correct docs~

Closes #98

### CODE REVIEW

File: src/nexgen/nxs_utils/Source.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+        self.probe = probe
+
+    def __repr__(self) -> str:
+        msg = f"Beamline {self.beamline} at {self.name}.\n"
+        if self.probe:
+            msg += "Probe: {}"
+        return msg
+
+    def set_instrument_name(self) -> str:
```
Comment (DominicOram) [inline src/nexgen/nxs_utils/Source.py:40]: This is never called in production and is a bit confusingly named given it doesn't set anything, just returns you what it calculates.

File: src/nexgen/nxs_utils/Source.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+        If source type is not defined a priori, the function will assume it is a Synchrotron.
+        If the facility_id is defined inside the source dictionary, that is the value that will be used.
+        Naming tries to follow the recommended convention for NXmx:
+        https://mmcif.wwpdb.org/dictionaries/mmcif_pdbx_v50.dic/Items/_diffrn_source.type.html
+
+        Returns:
+            name (str): The name to write under '/entry/instrument/name'
+        """
+        facility_id = "DIAMOND" if self.facility_id is None else self.facility_id
```
Comment (DominicOram) [inline src/nexgen/nxs_utils/Source.py:51]: Why not just set the default id to `DIAMOND` above?

File: src/nexgen/nxs_utils/Axes.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+# Define axes and scans
+@dataclass
+class Axis:
+    """
+    Define an axis object for goniometer or detector.
+    """
+
+    name: str
+    depends: str
+    transformation_type: Literal["translation", "rotation"]
```
Comment (DominicOram) [inline src/nexgen/nxs_utils/Axes.py:21]: It would be nicer if these were enums rather than strings

File: src/nexgen/nxs_utils/Axes.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+# Define axes and scans
+@dataclass
+class Axis:
+    """
+    Define an axis object for goniometer or detector.
+    """
+
+    name: str
+    depends: str
```
Comment (DominicOram) [inline src/nexgen/nxs_utils/Axes.py:20]: It still feels like this should be an `Axis`, we're basically creating a linked list

File: src/nexgen/nxs_utils/Detector.py [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+        beam_center: List[float],
+        exposure_time: float,
+        module_vectors: List[Point3D] | List[Tuple],
+    ):
+        if type(detector_params) not in [
+            EigerDetector,
+            TristanDetector,
+            SinglaDetector,
+        ]:
+            raise UnknownDetectorTypeError("Unknown detector.")
```
Comment (DominicOram) [inline src/nexgen/nxs_utils/Detector.py:112]: This should be covered by the type system, I don't think we need to check it twice

**Review by DominicOram — COMMENTED** [review]
I think we could do with an architectural review of this. When everyone is back lets get into a meeting room and spend a couple of hours going over the whole structure

File: src/nexgen/nxs_utils/Source.py (hunk 2 above)
Comment (noemifrisina) [reply src/nexgen/nxs_utils/Source.py:51]: Both this and the previous one are setting up for a couple of requests for Electron Diffraction metadata... refactoring that is still a wip.

**Pull request merged**